### DISCUSSION TITLE
Is there a way to specify a default/fallback transition for unexpected events? (@xstate/fsm)

### DISCUSSION THREAD

**Discussion by benblank** [statelyai/xstate#4057] [Q&A] [answered]
I'm just getting started with XState, using `@xstate/fsm` to track the state and context of a simple parser. In each state, there are a few events which correspond to the types of input expected, but I'd like any *un*expected input to put the parser into an error state.

I've been through the documentation, discussions, issues, and even the source, but so far the only way I know to accomplish this is to explicitly list a transition for *every event* on *every state*, most of which are the same:

``` javascript
const INVALID = "error";  // Just to make them more easily distinguished from the valid, string transitions.

const fsm = createMachine({
  initial: "preamble",
  states: {
    preamble: {
      on: {
        shortSeparator: "globalInfo",
        longSeparator: INVALID,
        blankLine: "preamble",
        property: INVALID,
        // …
        eof: INVALID,
      },
    },

    globalInfo: {
      on: {
        shortSeparator: INVALID,
        longSeparator: "entryList",
        blankLine: "globalInfo",
        property: "globalInfo",
        // …
        eof: INVALID,
      },
    },

    // ...
  },
});
```

Even with this simplified example, that's a fair amount of redundant-seeming configuration. I could make it a bit less verbose by creating an object which maps *all* events to the error state, then `{ on: Object.assign({}, allEventsInvalid, { event1: "stateA", event4: "stateG" }) }`, but what I was really hoping for was a way to change the default "if an event for which the current state has no specified transition arrives, *don't transition*" behavior (is that the same as "ignore the event"?) to "if an event for which the current state has no specified transition arrives, *transition to a default state*". In the case of my parser, that would be the error state.

This seems like a pretty natural behavior to want when managing state, so I was surprised when my searches and reading didn't turn up anything at all; it makes me worry that I simply failed in my search, rather than a solution not already existing. 😅

Is there a simple way to accomplish this with `@xstate/fsm`? (I didn't look into the full XState library as much, so I wouldn't be surprised if an elegant solution already exists there.) If not, would there be interest in a PR introducing such a feature? I imagine that, for example, it would be pretty straightforward to add a `fallbackTransition` property to the configuration which, if present, could be used [when no other is found](https://github.com/statelyai/xstate/blob/4909dab267840bf69b5025f48fccbd35faa12ba5/packages/xstate-fsm/src/index.ts#L189-L191).

**Comment by ChrisShank** [comment]
Xstate does have a wildcard trigger for these cases although im not sure if xstate/fsm supports that

https://xstate.js.org/docs/guides/transitions.html#wildcard-descriptors

**Comment by chaitanyaj2222** [comment, marked as answer]
@benblank, Below is the code with a `wildcard event descriptor` that is recommended by @ChrisShank 

```js
const INVALID = "error";  // Just to make them more easily distinguished from the valid, string transitions.

const fsm = createMachine({
  initial: "preamble",
  states: {
    preamble: {
      on: {
        shortSeparator: "globalInfo",
        "*": INVALID,
        blankLine: "preamble"
      },
    },

    globalInfo: {
      on: {
        "*": INVALID,
        longSeparator: "entryList",
        blankLine: "globalInfo",
        property: "globalInfo"
      },
    },

    // ...
  },
});
```

**Comment by benblank** [comment]
Thank you both! Unfortunately, it looks like `@xstate/fsm` doesn't support wildcards; asterisks get treated like any other event name.

``` javascript
const machine = createMachine({ initial: "start", states: { start: { on: { "*": "end" } }, end: {} } });

console.log(machine.transition(machine.initialState, "foo"));
// {
//   value: 'start',
//   context: undefined,
//   actions: [],
//   changed: false,
//   matches: [Function (anonymous)]
// }

console.log(machine.transition(machine.initialState, "*"));
// {
//   value: 'end',
//   context: undefined,
//   actions: [],
//   changed: true,
//   matches: [Function (anonymous)]
// }
```

I think I'll take a look at how wildcards are implemented in the full XState library and see whether making a PR adding them to `@xstate/fsm` makes sense.

**Reply by chaitanyaj2222** [reply]
I raised this [PR](https://github.com/statelyai/xstate/pull/4060). Based on whether this PR gets merged or denied, you would be able to decide on proceeding with your PR for adding `wildcard event descriptors` to `@xstate/fsm`.

edit: The above PR has been merged. So, I we can safely assume that `wildcard event descriptors` will not be a part of `@xstate/fsm`. To use `wildcard event descriptors` one must install `core xstate` package. Thanks!

**Reply by benblank** [reply]
Adding wildcards to @xstate/fsm turned out not to be bad at all. I created #4065 for it.

**Reply by chaitanyaj2222** [reply]
@benblank, I learnt a lot from you PR. Great work buddy.

**Reply by benblank** [reply]
Thanks! I learned a lot from it, too. 😉
